examples/improved_wgan.py

```diff
@@ -0,0 +1,263 @@
+    parser.add_argument("--output_dir", "-o", required=True, help="Directory to output generated files to")
+    args = parser.parse_args()
+
+    # First we load the image data, reshape it and normalize it to the range [-1, 1]
+    (X_train, y_train), (X_test, y_test) = mnist.load_data()
+    X_train = np.concatenate((X_train, X_test), axis=0)
+    X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], X_train.shape[2], 1))
+    X_train = (X_train.astype(np.float32) - 127.5) / 127.5
+
+    # Now we initialize the generator and discriminator.
+    generator = make_generator()
+    discriminator = make_discriminator()
+
+    # The generator_model is used when we want to train the generator layers.
+    # As such, we ensure that the discriminator layers are not trainable.
+    # Note that once we compile this model, updating .trainable will have no effect within it. As such, it
+    # won't cause problems if we later set discriminator.trainable = True for the discriminator_model, as long
+    # as we compile the generator_model first.
+    for layer in discriminator.layers:
+        layer.trainable = False
+    discriminator.trainable = False
+    generator_input = Input(shape=(100,))
+    generator_layers = generator(generator_input)
+    discriminator_layers_for_generator = discriminator(generator_layers)
+    generator_model = Model(inputs=[generator_input], outputs=[discriminator_layers_for_generator])
+    # We use the Adam paramaters from Gulrajani et al.
+    generator_model.compile(optimizer=Adam(0.0001, beta_1=0.5, beta_2=0.9), loss=wasserstein_loss)
+
+    # Now that the generator_model is compiled, we can make the discriminator layers trainable.
+    for layer in discriminator.layers:
+        layer.trainable = True
+    for layer in generator.layers:
+        layer.trainable = False
+    discriminator.trainable = True
+    generator.trainable = False
+
+    # The discriminator_model is more complex. It takes both real image samples and random noise seeds as input.
+    # The noise seed is run through the generator model to get generated images. Both real and generated images
+    # are then run through the discriminator. Although we could concatenate the real and generated images into a
+    # single tensor, we don't (see model compilation for why).
+    real_samples = Input(shape=X_train.shape[1:])
+    generator_input_for_discriminator = Input(shape=(100,))
+    generated_samples_for_discriminator = generator(generator_input_for_discriminator)
+    discriminator_output_from_generator = discriminator(generated_samples_for_discriminator)
+    discriminator_output_from_real_samples = discriminator(real_samples)
+
+    # We also need to generate weighted-averages of real and generated samples, to use for the gradient norm penalty.
+    averaged_samples = RandomWeightedAverage()([real_samples, generated_samples_for_discriminator])

```

Comment:
Can you explain what you are doing here and why? I don't see anything like this in the paper. It seems like you are averaging the pixels of generated samples and real samples and then feeding to the discriminator. This doesn't seem right at all.